editor/src/components/canvas/ui-jsx-canvas.tsx

```diff
@@ -597,162 +598,171 @@ export function attemptToResolveParsedComponents(
   filePathResolveResult: Either<string, string>,
   editedText: ElementPath | null,
 ): Either<string, MapLike<any>> {
-  return flatMapEither((resolvedFilePath) => {
-    resolvedFromThisOrigin.push(toImport)
-    const projectFile = getProjectFileByFilePath(projectContents, resolvedFilePath)
-    if (
-      projectFile != null &&
-      isTextFile(projectFile) &&
-      isParseSuccess(projectFile.fileContents.parsed)
-    ) {
-      const exportsDetail = projectFile.fileContents.parsed.exportsDetail
-      // Should only use the full scope and components support if the file contains components
-      // or if it does any kind of re-exporting as we can't guarantee that the re-exported
-      // files do not contain components.
-      // Exclude any file with the destructured assignment export style as they're quite problematic
-      // to support.
-      const shouldUseFileScope =
-        (projectFile.fileContents.parsed.topLevelElements.some(isUtopiaJSXComponent) ||
-          exportsDetail.some(isReexportExportDetail)) &&
-        !exportsDetail.some(isExportDestructuredAssignment)
-      if (shouldUseFileScope) {
-        const { scope } = createExecutionScope(
-          resolvedFilePath,
-          customRequire,
-          mutableContextRef,
-          topLevelComponentRendererComponents,
-          projectContents,
-          uiFilePath,
-          base64FileBlobs,
-          hiddenInstances,
-          displayNoneInstances,
-          metadataContext,
-          updateInvalidatedPaths,
-          shouldIncludeCanvasRootInTheSpy,
-          editedText,
-        )
-        let filteredScope: MapLike<any> = {
-          ...scope.module.exports,
-          __esModule: true,
-        }
+  return foldEither(
+    (msg) => {
+      // This import in this file has already been resolved, so we check for an existing scope
+      const existingScope = resolvedFromThisOrigin[toImport]
+      return existingScope == null ? left(msg) : right(existingScope)
+    },
+    (resolvedFilePath) => {
+      resolvedFromThisOrigin[toImport] = null // We use null as a marker to indicate that we have started resolving, but not completed yet
```

Comment:
:+1

Reply:
can't we use some other marker like `'started-resolving'`? I think the `!== undefined` (from above) is very unexpressive, and the shape of the code would stay the same